src/env/node/git/localGitProvider.ts

```diff
@@ -1252,13 +1253,87 @@ export class LocalGitProvider implements GitProvider, Disposable {
 	}
 
 	@log()
-	async createBranch(repoPath: string, name: string, ref: string): Promise<void> {
-		await this.git.branch(repoPath, name, ref);
+	createBranch(repoPath: string, name: string, ref: string): Promise<void> {
+		try {
+			return void this.git.branch(repoPath, name, ref);
+		} catch (ex) {
+			if (ex instanceof BranchError) {
+				throw ex.WithBranch(branch.name);
+			}
+
+			throw ex;
+		}
+	}
+
+	@log()
+	renameBranch(repoPath: string, oldName: string, newName: string): Promise<void> {
+		try {
+			return void this.git.branch(repoPath, '-m', oldName, newName);
+		} catch (ex) {
+			if (ex instanceof BranchError) {
+				throw ex.WithBranch(branch.name);
+			}
+
+			throw ex;
+		}
 	}
 
 	@log()
-	async renameBranch(repoPath: string, oldName: string, newName: string): Promise<void> {
-		await this.git.branch(repoPath, '-m', oldName, newName);
+	async deleteBranch(
+		repoPath: string,
+		branch: GitBranchReference,
+		options: { force?: boolean; remote?: boolean },
+	): Promise<void> {
+		try {
+			if (branch.remote) {
+				await this.git.push(repoPath, {
+					delete: {
+						remote: getRemoteNameFromBranchName(branch.name),
+						branch: branch.remote ? getBranchNameWithoutRemote(branch.name) : branch.name,
+					},
+				});
+				return;
+			}
+
+			const args = ['--delete'];
+			if (options.force) {
+				args.push('--force');
+			}
+
+			if (!options.remote || !branch.upstream) {
+				await this.git.branch(repoPath, ...args, branch.ref);
+				return;
+			}
+
+			const remote = getRemoteNameFromBranchName(branch.upstream.name);
+			remoteCommit = await this.git.rev_list(repoPath, `refs/remotes/${remote}/${branch.ref}`, {
+				maxResults: 1,
+			});
+
+			await this.git.branch(repoPath, '--delete', '--remotes', `${remote}/${branch.ref}`);
+
+			try {
+				await this.git.branch(repoPath, ...args, branch.ref);
+			} catch (ex) {
+				// If it fails, restore the remote branch
+				await this.git.update_ref(repoPath, `refs/remotes/${remote}/${branch.ref}`, commit);
+				await this.git.branch__set_upstream(repoPath, branch, remote, branch);
```

Comment:
`commit` doesn't exist in this context.

`branch` is a `GitBranchReference` and `branch__set_upstream` is looking for a string